standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java

```diff
@@ -609,22 +609,10 @@ public void init() throws MetaException {
         partitionValidationPattern = null;
       }
 
-      // We only initialize once the tasks that need to be run periodically
-      if (alwaysThreadsInitialized.compareAndSet(false, true)) {
-        ThreadPool.initialize(conf);
-        Collection<String> taskNames =
-            MetastoreConf.getStringCollection(conf, ConfVars.TASK_THREADS_ALWAYS);
-        for (String taskName : taskNames) {
-          MetastoreTaskThread task =
-              JavaUtils.newInstance(JavaUtils.getClass(taskName, MetastoreTaskThread.class));
-          task.setConf(conf);
-          long freq = task.runFrequency(TimeUnit.MILLISECONDS);
-          // For backwards compatibility, since some threads used to be hard coded but only run if
-          // frequency was > 0
-          if (freq > 0) {
-            ThreadPool.getPool().scheduleAtFixedRate(task, freq, freq, TimeUnit.MILLISECONDS);
-          }
-        }
+      // We only initialize once the tasks that need to be run periodically. For remote metastore
+      // these threads are started along with the other housekeeping threads only in the leader HMS.
+      if (!isMetaStoreRemote()) {
```

Comment:
Even with this change, they will be run in embedded mode also. *!* isMetaStoreRemote() takes care of that. !isMetaStoreRemote() implies embedded mode.

Reply:
If config is not set, it will be empty, and so the HMS will be leader, so the tasks will be run. If the config is set to some other host, the "always threads" won't be started since it's not a leader. But in an embedded metastore where this config is set, we shouldn't start the threads since the leader is going to run those. I have made this change and uploaded a new patch.